README.md
# sql-backup - Backup your MySQL / MariaDB server !

[![Codacy Badge](https://app.codacy.com/project/badge/Grade/1d6a522144ca4169a0c679bd9d299341)](https://www.codacy.com/gh/williamdes/sql-backup/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=williamdes/sql-backup&amp;utm_campaign=Badge_Grade)
[![Actions Status](https://github.com/williamdes/sql-backup/workflows/Run%20tests/badge.svg)](https://github.com/williamdes/sql-backup/actions)
[![codecov](https://codecov.io/gh/williamdes/sql-backup/branch/master/graph/badge.svg)](https://codecov.io/gh/williamdes/sql-backup)
[![License: Unlicense](https://img.shields.io/badge/license-Unlicense-blue.svg)](http://unlicense.org/)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fwilliamdes%2Fsql-backup.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2Fwilliamdes%2Fsql-backup?ref=badge_shield)
[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/1827/badge)](https://bestpractices.coreinfrastructure.org/projects/1827)

Backup your MySQL / MariaDB server ( structure, data, users, grants, views, triggers, routines, events )

## Install

### Clone the repository

```sh
git clone https://github.com/williamdes/sql-backup.git --depth 1
```

### Move into the directory

```sh
cd sql-backup
```

### Copy the configuration example

```sh
cp .env.dist .env
```

### Edit your configuration file

#### Nano

```sh
nano .env
```

#### Vi

```sh
vi .env
```

### Start backup

Execute this command to start the backup !

```sh
./backup.sh
```

## Extras

To use an external env file in a custom location, example :

```sh

export BACKUP_CONFIG_ENVFILE="/home/myuser/secretbackupconfig.env.txt"

./backup.sh

unset BACKUP_CONFIG_ENVFILE
```

You can use the variables of the env file in the `ON_SUCCESS` variable, example :

```sh
ON_SUCCESS="${BACKUP_DIR}/onsuccessscript.sh"
```

## Options

| NAME           | DEFAULT VALUE                                                                                                                  | OPTIONAL |
|----------------|--------------------------------------------------------------------------------------------------------------------------------|----------|
| BACKUP_DIR     |                                                                                                                                | NO       |
| MYSQL_HOST     |                                                                                                                                | NO       |
| MYSQL_USER     |                                                                                                                                | NO       |
| MYSQL_PASS     |                                                                                                                                | YES      |
| MYSQL_PORT     | 3306                                                                                                                           | YES      |
| SKIP_DATABASES |                                                                                                                                | YES      |
| EXPERT_ARGS    | --default-character-set=utf8 --extended-insert=FALSE --single-transaction --skip-comments --skip-dump-date --hex-blob --tz-utc | YES      |
| ON_SUCCESS     |                                                                                                                                | YES      |
| SKIP_OP        |                                                                                                                                | YES      |

> ON_SUCCESS is called on script success

## Example .env

```sh
BACKUP_DIR=/sql_backup
MYSQL_HOST=localhost
MYSQL_USER=root
MYSQL_PASS=root
MYSQL_PORT=3306
SKIP_DATABASES=mysql,sys,information_schema,performance_schema,phpmyadmin
SKIP_OP=users,grants
```

## Files

| NAME          | DESCRIPTION                               |
|---------------|-------------------------------------------|
| database.sql  | All the data                              |
| structure.sql | The structure of the databases and tables |
| grants.sql    | The grants for all users except root      |
| events.sql    | The scheduled events                      |
| views.sql     | The views                                 |
| triggers.sql  | The triggers                              |
| routines.sql  | All the procedures & functions            |
| users.sql     | All MySQL users                           |

## License

[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fwilliamdes%2Fsql-backup.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2Fwilliamdes%2Fsql-backup?ref=badge_large)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2020-12-08

- Added a default value for MYSQL_PORT

## [1.1.0] - 2019-12-01

### Added

- Check if mysql and mysqldump are installed
- Updated README.md
- Added optional `BACKUP_CONFIG_ENVFILE` variable to set the env file location
- Updated .env.dist (commented out some optional variables)
- Added check if `BACKUP_DIR` exists
- Added exit codes
- Added unit tests
- Updated .gitignore
- Bug fix (no views)
- Added option `SKIP_OP`
- Added option `MYSQL_PORT`
- Migrate from TravisCI to GitHub actions
- Some README improvements

## [1.0.0] - 2018-04-07

### Added

- CHANGELOG
- README
- LICENCE
- .env.dist
- .gitignore
- backup.sh


tests.sh
#!/bin/bash
# Error codes
# 200 = Empty Variable
# 201 = Value is not a file
# 202 = Value is not a directory
# 203 = Missing dependency
# 204 = Databases listing failed
# 205 = Views listing failed
# 206 = No databases to backup
# 207 = Structure dump failed
# 208 = Data dump failed
# 209 = Routines dump failed
# 210 = Triggers dump failed
# 211 = Events dump failed
# 212 = Views dump failed
# 213 = Users dump failed
# 214 = Grants dump failed
# 215 = Grants list failed

TEST_MYSQL_HOST="${TEST_MYSQL_HOST:-localhost}"
TEST_MYSQL_USER="${TEST_MYSQL_USER:-root}"
TEST_MYSQL_PASS="${TEST_MYSQL_PASS:-}"
TEST_MYSQL_PORT="${TEST_MYSQL_PORT:-3306}"
SCRIPT_ROOT="$(dirname $0)"
echo "SCRIPT_ROOT=${SCRIPT_ROOT}"
MYSQLCREDS="${MYSQLCREDS:---host ${TEST_MYSQL_HOST} -u${TEST_MYSQL_USER} -p${TEST_MYSQL_PASS} --port ${TEST_MYSQL_PORT}}"

compareFiles() {
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/structure.sql" "${SCRIPT_ROOT}/test/structure.sql"
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/database.sql" "${SCRIPT_ROOT}/test/database.sql"
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/grants.sql" "${SCRIPT_ROOT}/test/grants.sql"
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/users.sql" "${SCRIPT_ROOT}/test/users.sql"
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/events.sql" "${SCRIPT_ROOT}/test/events.sql"
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/views.sql" "${SCRIPT_ROOT}/test/views.sql"
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/triggers.sql" "${SCRIPT_ROOT}/test/triggers.sql"
  compareFilesOrExit "${SCRIPT_ROOT}/samples/$1/routines.sql" "${SCRIPT_ROOT}/test/routines.sql"
}

sourceCopy() {
  sed -n "/^$1()/,/^}/p" ./backup.sh > func.sh; source func.sh; rm func.sh
}

compareFilesSUM() {

  chk1="$(sha1sum $1 | awk -F" " '{print $1}')"
  chk2="$(sha1sum $2 | awk -F" " '{print $1}')"

  if [ "$chk1" = "$chk2" ]; then
    return 0
  else
    return 1
  fi

}

compareFilesOrExit() {
  diff -ia --unified=1 --suppress-common-lines "$1" "$2"
  compareFilesSUM "$@" || fail "Files are not identical ($1) ($2)"
}

testEMPTY_VAR_Fail() {
  ./backup.sh 1>/dev/null 2>&1
  assertEquals 200 "$?"
}

testBACKUP_CONFIG_ENVFILE_Fail() {
  export BACKUP_CONFIG_ENVFILE="/home/myuser/secretbackupconfig.env.txt"
  ./backup.sh
  assertEquals 201 "$?"
  unset BACKUP_CONFIG_ENVFILE
}

fillConfigFile() {
  echo "MYSQL_HOST=${TEST_MYSQL_HOST}" >> $1
  echo "MYSQL_USER=${TEST_MYSQL_USER}" >> $1
  echo "MYSQL_PASS=${TEST_MYSQL_PASS}" >> $1
  echo "MYSQL_PORT=${TEST_MYSQL_PORT}" >> $1
  echo "SKIP_DATABASES=mysql,sys,information_schema,performance_schema" >> $1
}

createTestData() {
  mysql ${MYSQLCREDS} < ${SCRIPT_ROOT}/samples/$1/create.sql
}

destroyTestData() {
  mysql ${MYSQLCREDS} < ${SCRIPT_ROOT}/samples/$1/destroy.sql
}

preTest() {
  mkdir ./test
  touch ./test/envfile
  echo "BACKUP_DIR=${SCRIPT_ROOT}/test" > ./test/envfile
  fillConfigFile ./test/envfile
  export BACKUP_CONFIG_ENVFILE="./test/envfile"
}

postTest() {
  rm ./test/*
  rmdir ./test
  unset BACKUP_CONFIG_ENVFILE
}

testBACKUP_CONFIG_ENVFILE_Success() {
  preTest
  echo "MYSQL_HOST=tests.test" >> "./test/envfile"
  ./backup.sh
  assertEquals 204 "$?"
  postTest
}

testBACKUP_EMPTY_Success() {
  preTest
  ./backup.sh
  assertEquals 206 "$?"
  postTest
}

testBACKUP_Success_NoDiff() {
  preTest
  createTestData "empty"
  ./backup.sh
  assertEquals 0 "$?"
  destroyTestData "empty"
  compareFiles "empty"
  postTest
}

testBACKUP_Success() {
  preTest
  createTestData "empty"
  ./backup.sh
  assertEquals 0 "$?"
  destroyTestData "empty"
  postTest
}

testOnSuccessScript() {
  preTest
  echo 'ON_SUCCESS="./postTest.sh"' >> "./test/envfile"
  createTestData "empty"
  ./backup.sh
  cat ./test/endfile > /dev/null
  assertEquals 0 "$?"
  destroyTestData "empty"
  postTest
}

testCompareFail() {
  # Expected fail
  compareFilesSUM "${SCRIPT_ROOT}/.gitignore" "${SCRIPT_ROOT}/samples/empty/events.sql"
  assertEquals 1 "$?"
}

testBACKUP_Success_NoDiff_Strange_BS_Data0() {
  preTest
  createTestData "withdata0"
  ./backup.sh
  assertEquals 0 "$?"
  destroyTestData "withdata0"
  compareFiles "withdata0"
  postTest
}

testBACKUP_Success_NoDiff_Strange_BS_Data1() {
  preTest
  createTestData "withdata1"
  ./backup.sh
  assertEquals 0 "$?"
  destroyTestData "withdata1"
  compareFiles "withdata1"
  postTest
}

testBACKUP_Success_NoDiff_Strange_BS_Data2() {
  preTest
  createTestData "withdata2"
  ./backup.sh
  assertEquals 0 "$?"
  destroyTestData "withdata2"
  compareFiles "withdata2"
  postTest
}

testBACKUP_Success_NoDatabases() {
  preTest
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  ./backup.sh
  assertEquals 206 "$?"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testEvents_NoData_Fail() {
  preTest
  createTestData "withdata0"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT SELECT ON testbench.* TO 'grantfail'@'%';"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  ./backup.sh
  assertEquals 211 "$?"
  destroyTestData "withdata0"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testViewsDump_Fail() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT INSERT ON testbench.* TO 'grantfail'@'%';"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  echo "SKIP_OP=structure,data,routines,events,triggers,users,grants" >> "./test/envfile"
  ./backup.sh
  assertEquals 212 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testUsers_Fail() {
  preTest
  createTestData "withdata0"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToTestBench.sql"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  ./backup.sh
  assertEquals 213 "$?"
  destroyTestData "withdata0"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testEvents_Fail() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT SELECT, TRIGGER, SHOW VIEW ON testbench.* TO 'grantfail'@'%';"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  ./backup.sh
  assertEquals 211 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testStructure_Fail() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT INSERT, TRIGGER, SHOW VIEW ON testbench.* TO 'grantfail'@'%';"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  ./backup.sh
  assertEquals 207 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testSkip_HasNot() {
  export SKIP_OP=()
  sourceCopy "hasSkip"
  hasSkip "t"
  assertEquals 0 "$?"
  unset SKIP_OP
}

testSkip_Has() {
  export SKIP_OP=("skip1")
  sourceCopy "hasSkip"
  hasSkip "skip1"
  assertEquals 1 "$?"
  unset SKIP_OP
}

testData_Fail() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT INSERT ON testbench.* TO 'grantfail'@'%';"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  echo "SKIP_OP=structure,routines,triggers,events,views,users,grants" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  ./backup.sh
  assertEquals 208 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testRoutines_Fail() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT INSERT ON testbench.* TO 'grantfail'@'%';"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/withdata2/function1.sql"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  echo "SKIP_OP=structure,data,triggers,events,views,users,grants" >> "./test/envfile"
  ./backup.sh
  assertEquals 209 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testTriggers_Fail() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT INSERT ON testbench.* TO 'grantfail'@'%';"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/withdata2/trigger1.sql"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  echo "SKIP_OP=structure,data,routines,events,views,users,grants" >> "./test/envfile"
  ./backup.sh
  assertEquals 210 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testGrantsList_Fail() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/withdata2/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT INSERT ON testbench.* TO 'userwd2gf'@'%';"
  echo "MYSQL_USER=userwd2gf" >> "./test/envfile"
  echo "SKIP_OP=structure,data,routines,events,triggers,users,views" >> "./test/envfile"
  ./backup.sh
  assertEquals 215 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/withdata2/deleteuser.sql"
  postTest
}

#testGrants_Fail() {
#  preTest
#  createTestData "withdata2"
#  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/withdata2/createuser.sql"
#  mysql ${MYSQLCREDS} -ANe "GRANT INSERT ON testbench.* TO 'userwd2gf'@'%';"
#  echo "MYSQL_USER=userwd2gf" >> "./test/envfile"
#  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/withdata2/grantToUsers.sql"
#  echo "SKIP_OP=structure,data,routines,events,triggers,users,views" >> "./test/envfile"
#  ./backup.sh
#  assertEquals 214 "$?"
#  destroyTestData "withdata2"
#  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/withdata2/deleteuser.sql"
#  postTest
#}

testBACKUP_manualgrant_Success() {
  preTest
  createTestData "withdata0"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToTestBench.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT ALL PRIVILEGES ON mysql.* TO 'grantfail'@'%';"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  ./backup.sh
  assertEquals 0 "$?"
  destroyTestData "withdata0"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}

testNoSkipDatabases() {
  preTest
  createTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/createuser.sql"
  mysql ${MYSQLCREDS} -ANe "GRANT INSERT ON testbench.* TO 'grantfail'@'%';"
  echo "MYSQL_USER=grantfail" >> "./test/envfile"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToUsers.sql"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/grantToDB.sql"
  echo "SKIP_OP=structure,data,routines,events,triggers,users,grants" >> "./test/envfile"
  echo "SKIP_DATABASES=" >> "./test/envfile"
  ./backup.sh
  assertEquals 0 "$?"
  destroyTestData "withdata2"
  mysql ${MYSQLCREDS} < "${SCRIPT_ROOT}/samples/empty/deleteuser.sql"
  postTest
}
export SHUNIT_COLOR="none" # GitHub actions
. ./shunit2-2.1.8/shunit2

backup.sh
#!/bin/bash

exitWithMsg() {
  echo $2
  exit $1
}

hasSkip() {
  for e in "${SKIP_OP[@]}"; do
    if [ "$e" == "$1" ]; then
      return 1
    fi
  done
  return 0
}

# from : github:builtinnya/dotenv-shell-loader
DOTENV_SHELL_LOADER_SAVED_OPTS=$(set +o)
set -o allexport
if [ ! -z "${BACKUP_CONFIG_ENVFILE}" ]; then
    if [ ! -f "${BACKUP_CONFIG_ENVFILE}" ]; then
        exitWithMsg 201 "Value of variable BACKUP_CONFIG_ENVFILE is not a file (${BACKUP_CONFIG_ENVFILE})"
    else
      [ -f "${BACKUP_CONFIG_ENVFILE}" ] && source "${BACKUP_CONFIG_ENVFILE}"
    fi
else
  [ -f "$(dirname $0)/.env" ] && source "$(dirname $0)/.env"
fi
set +o allexport
eval "$DOTENV_SHELL_LOADER_SAVED_OPTS"
unset DOTENV_SHELL_LOADER_SAVED_OPTS

if ! [ -x "$(command -v mysql)" ]; then
  exitWithMsg 203 'Error: mysql is not installed !, apt-get install -y mysql-client' >&2
fi

if ! [ -x "$(command -v mysqldump)" ]; then
  exitWithMsg 203 'Error: mysqldump is not installed !, apt-get install -y mysql-client' >&2
fi

if [ -z "${BACKUP_DIR}" ]; then
    exitWithMsg 200 "Empty Variable BACKUP_DIR"
else
  if [ ! -d "${BACKUP_DIR}" ]; then
      exitWithMsg 202 "Value of variable BACKUP_DIR is not a directory (${BACKUP_DIR})"
  fi
fi

if [ -z "${MYSQL_HOST}" ]; then
    exitWithMsg 200 "Empty Variable MYSQL_HOST"
fi

if [ -z "${MYSQL_USER}" ]; then
    exitWithMsg 200 "Empty Variable MYSQL_USER"
fi

if [ -z "${SKIP_OP}" ]; then
    SKIP_OP=()
else
  SKIP_OPS=$(echo -e "${SKIP_OP}" | tr "," "\n")
  SKIP_OP=()
  oldIFS=$IFS
  IFS=$'\n'
  for SKIP in $SKIP_OPS; do
    SKIP_OP+=(${SKIP} )
  done
  IFS=$oldIFS
fi

MYSQL_CONN="-h${MYSQL_HOST} -u${MYSQL_USER} --port ${MYSQL_PORT:-3306}"

if [ ! -z "${MYSQL_PASS}" ]; then
    MYSQL_CONN="${MYSQL_CONN} -p${MYSQL_PASS} --port ${MYSQL_PORT:-3306}"
fi

if [ -z "${EXPERT_ARGS}" ]; then
    EXPERT_ARGS="--default-character-set=utf8 --extended-insert=FALSE --single-transaction --skip-comments --skip-dump-date --hex-blob --tz-utc"
fi

MYSQLDUMP_DEFAULTS="${MYSQL_CONN} ${EXPERT_ARGS}"

DB_LIST_SQL="SELECT schema_name FROM information_schema.schemata"

# If ${SKIP_DATABASES} is not empty, create a where chain
if [ ! -z "${SKIP_DATABASES}" ]; then
    DB_LIST_SQL="${DB_LIST_SQL} WHERE schema_name NOT IN ("
    # Split on ,
    SKIP_DATABASES=$(echo -e "${SKIP_DATABASES}" | tr "," "\n")
    for DB in ${SKIP_DATABASES} ; do
    DB_LIST_SQL="${DB_LIST_SQL}'${DB}'," ;
    done
    DB_LIST_SQL="${DB_LIST_SQL: : -1}"
    DB_LIST_SQL="${DB_LIST_SQL});"
fi
#    echo -e "${DB_LIST_SQL}"
# Get result
DB_LIST=$(mysql ${MYSQL_CONN} -ANe"${DB_LIST_SQL}")

if [ "$?" -ne 0 ]; then
  exitWithMsg 204 "Databases listing failed"
fi

if [ -z "${DB_LIST}" ]; then
  exitWithMsg 206 "No databases to backup"
fi

for DB in $DB_LIST; do # Concat ignore command
    DBS="${DBS} ${DB}"
done

hasSkip "views"
if [ "$?" -eq 0 ]; then
  VIEW_LIST_SQL="SET SESSION group_concat_max_len = 1000000;SELECT IFNULL(GROUP_CONCAT(concat(':!\`',table_schema,'\`.\`',table_name,'\`') SEPARATOR ''),'') FROM information_schema.views"

  # If ${SKIP_DATABASES} is not empty, create a where chain
  if [ ! -z "${SKIP_DATABASES}" ]; then
      VIEW_LIST_SQL="${VIEW_LIST_SQL} WHERE table_schema NOT IN ("
      # Split on ,
      SKIP_DATABASES=$(echo -e "${SKIP_DATABASES}" | tr "," "\n")
      for DB in ${SKIP_DATABASES} ; do
        VIEW_LIST_SQL="${VIEW_LIST_SQL}'${DB}'," ;
      done
      VIEW_LIST_SQL="${VIEW_LIST_SQL: : -1}"
      VIEW_LIST_SQL="${VIEW_LIST_SQL});"
  else
      VIEW_LIST_SQL=";"
  fi

  # Get result
  VIEWS_LIST=$(mysql ${MYSQL_CONN} -ANe"${VIEW_LIST_SQL}")

  if [ "$?" -ne 0 ]; then
    exitWithMsg 205 "Views listing failed"
  fi

  VIEW_IGNORE_ARG=()
  # Split on :!
  VIEWS=$(echo -e "${VIEWS_LIST}" | tr ":!" "\n")
  # echo -e "${VIEWS}"

  oldIFS=$IFS
  IFS=$'\n'
  for VIEW in $VIEWS; do # Concat ignore command
    # Replace ` in ${VIEW}, does not work with ` for --ignore-table
    VIEW="${VIEW//\`/}"
    #VIEW=$(printf '%q' "${VIEW}")
    VIEW_IGNORE_ARG+=(--ignore-table=${VIEW} )
  done
  IFS=$oldIFS
  # echo "${VIEW_IGNORE_ARG[@]}";
fi

hasSkip "structure"
if [ "$?" -eq 0 ]; then
  echo "Structure..."
  mysqldump ${MYSQLDUMP_DEFAULTS} --skip-add-drop-table --routines=FALSE --triggers=FALSE --events=FALSE --no-data "${VIEW_IGNORE_ARG[@]}" --databases ${DBS} > ${BACKUP_DIR}/structure.sql

  if [ "$?" -ne 0 ]; then
    exitWithMsg 207 "Structure dump failed"
  fi
fi

hasSkip "data"
if [ "$?" -eq 0 ]; then
  echo "Data ..."
  mysqldump ${MYSQLDUMP_DEFAULTS} --routines=FALSE --triggers=FALSE --events=FALSE --no-create-info "${VIEW_IGNORE_ARG[@]}" --databases ${DBS} > ${BACKUP_DIR}/database.sql

  if [ "$?" -ne 0 ]; then
    exitWithMsg 208 "Data dump failed"
  fi
fi

hasSkip "routines"
if [ "$?" -eq 0 ]; then
  echo "Routines ..."
  mysqldump ${MYSQLDUMP_DEFAULTS} --routines=TRUE --triggers=FALSE --events=FALSE --no-create-info --no-data --no-create-db --databases ${DBS} > ${BACKUP_DIR}/routines.sql

  if [ "$?" -ne 0 ]; then
    exitWithMsg 209 "Routines dump failed"
  fi
fi

hasSkip "triggers"
if [ "$?" -eq 0 ]; then
  echo "Triggers ..."
  mysqldump ${MYSQLDUMP_DEFAULTS} --routines=FALSE --triggers=TRUE --events=FALSE --no-create-info --no-data --no-create-db --databases ${DBS} > ${BACKUP_DIR}/triggers.sql

  if [ "$?" -ne 0 ]; then
    exitWithMsg 210 "Triggers dump failed"
  fi
fi

hasSkip "events"
  if [ "$?" -eq 0 ]; then
  echo "Events ..."
  mysqldump ${MYSQLDUMP_DEFAULTS} --routines=FALSE --triggers=FALSE --events=TRUE --no-create-info --no-data --no-create-db --databases ${DBS} > ${BACKUP_DIR}/events.sql

  if [ "$?" -ne 0 ]; then
    exitWithMsg 211 "Events dump failed"
  fi
fi

hasSkip "views"
if [ "$?" -eq 0 ]; then
  echo "Views ..."
  VIEWS_SHOW_SQL=""
  oldIFS=$IFS
  IFS=$'\n'
  for VIEW in $VIEWS; do # Concat SHOW CREATE VIEW command
      VIEWS_SHOW_SQL="${VIEWS_SHOW_SQL}SHOW CREATE VIEW ${VIEW};"
  done
  IFS=$oldIFS

  # echo -e "${VIEWS_SHOW_SQL}"
  echo ${VIEWS_SHOW_SQL} | sed 's/;/\\G/g' | mysql ${MYSQL_CONN} > ${BACKUP_DIR}/views.sql

  if [ "$?" -ne 0 ]; then
    exitWithMsg 212 "Views dump failed"
  fi

  #Keeps lines starting with Create
  sed -i '/Create/!d' ${BACKUP_DIR}/views.sql
  # Removes 'Create View'
  sed -i -e 's/Create\ View://g' ${BACKUP_DIR}/views.sql
  #add ; at lines end
  sed -i 's/)$/);/' ${BACKUP_DIR}/views.sql
  #Remove spaces before start of line
  sed -i 's/^ *//' ${BACKUP_DIR}/views.sql
  #Add ; at line end
  sed -i 's/$/;/' ${BACKUP_DIR}/views.sql
  #Replace double ;; by ;
  sed -i 's/;;/;/' ${BACKUP_DIR}/views.sql
fi

hasSkip "users"
if [ "$?" -eq 0 ]; then
  echo "Users ..."
  mysqldump ${MYSQLDUMP_DEFAULTS} mysql --no-create-info --complete-insert --tables user db > ${BACKUP_DIR}/users.sql

  if [ "$?" -ne 0 ]; then
    exitWithMsg 213 "Users dump failed"
  fi
fi

hasSkip "grants"
if [ "$?" -eq 0 ]; then
  echo "Grants ..."
  # Needs refactor
  GRANTS_SQL="select distinct concat( \"SHOW GRANTS FOR '\",user,\"'@'\",host,\"';\" ) from mysql.user WHERE user != 'root';"
  GRANTS_LIST=$(mysql ${MYSQL_CONN} -ANe"${GRANTS_SQL}")
  if [ "$?" -ne 0 ]; then
    exitWithMsg 215 "Grants list failed"
  fi
  GRANTS_LIST=$(echo ${GRANTS_LIST} | mysql --default-character-set=utf8 --skip-comments ${MYSQL_CONN})
  if [ "$?" -ne 0 ]; then
    exitWithMsg 214 "Grants dump failed"
  fi
  echo ${GRANTS_LIST} | sed 's/\(GRANT .*\)/\1;/;s/^\(Grants for .*\)/-- \1 --/;/--/{x;p;x;}' > ${BACKUP_DIR}/grants.sql


  # Removes double backslashes >  \\
  sed -i -e 's/\\\\//g' ${BACKUP_DIR}/grants.sql
  # echo -e ${GRANTS_SQL}
fi


echo "Backup done !"

if [ ! -z "${ON_SUCCESS}" ]; then
  echo "$(${ON_SUCCESS})"
fi
exit 0